feat: chat scaffold

brianp opened the pull request:
Description
---
- Refactor the contacts crate for better organization.
- Expand contacts with types, and capabilities for managing messaging over the existing comms network
- Add cucumber tests to provide support for new functionality

Motivation and Context
---
We want to expand Tari's communication network to handle p2p chat services. This PR sets us off in the right direction for the core Tari processes to be able to handle message passing and storage. Currently chat clients will be able to send, and receive messages when online or offline. Utilizing store and forward for sending communications offline.

How Has This Been Tested?
---
Via cucumber only. The chat client within cucumber is currently the only interface for trials.

Caveats and known issues
---
- [ ] Encryption at rest: We need to change the message storage to support encryption at rest

brianp committed:
Expand contacts to handle messaging

This refactor reorganizes the contacts crate and builds out the
possibility of sending messages to contacts via the already running
comms services. Messages will be monitored for and stored locally for
the client to retrieve and present in whatever form.

brianp committed:
Fixup lingering references to the contacts crate

These references are out of date after the refactor and just needed a
little guidance.

brianp committed:
Add cucumber tests for chat

This provides a barebones chat client used for testing within cucumber.
The client is relatively isolated from the test suite itself and could
be expanded on in the future.

brianp committed:
Clippy fixes and function updating

brianp committed:
Correct the timestamps so it doesn't put us years into the future

brianp committed:
Remove cucumber waits and add timeout loops

brianp committed:
Use more encyrption (more then none)

I was using the default encryption method, which happens to be
plaintext. This caused all messages to go out unencrypted and allowed
all clients to receive and read all messages. With the appropriate
encryption methods client can't read messages not for them, and ignore
them, or SAF.

brianp committed:
Remove client LocalNet dependency

brianp committed:
Use the authenticated origin, not the source

brianp committed:
Get direct sends working

brianp committed:
Make sure we be logging good like

Update the log4s to include the refactoring of contacts and surrounding
services.

hansieodendaal commented on integration_tests/tests/features/Chat.feature:
```diff
@@ -0,0 +1,43 @@
+
+Feature: Chat messaging
+
+  Scenario: A message is propagated between nodes
```
```suggestion
  Scenario: A message is propagated between nodes via a 3rd party base node
```

hansieodendaal commented on integration_tests/tests/features/Chat.feature:
```diff
@@ -0,0 +1,43 @@
+    Then CHAT_B will have 1 message with CHAT_A
+
+  Scenario: A message is sent directly between nodes
+    Given I have a seed node SEED_A
+    When I have a chat client CHAT_A connected to seed node SEED_A
+    When I have a chat client CHAT_B connected to seed node SEED_A
```
Is it still necessary to have the 3rd party base node here?

brianp replied:
Without the seed node the two chat clients don't seem to find each other after being added as contacts. We would probably need to introduce a peer dialing step for new contacts if the client doesn't otherwise have connectivity.

hansieodendaal replied:
Maybe we can register an issue to monitor if this will be a problem for the mobile wallets, as they will most probably not connect to the same base node at startup.

hansieodendaal commented on integration_tests/tests/features/Chat.feature:
```diff
@@ -0,0 +1,43 @@
+    When I have a chat client CHAT_A connected to seed node SEED_A
+    When I have a chat client CHAT_B connected to seed node SEED_A
+    When I add CHAT_B as a contact to CHAT_A
+    When CHAT_A waits for contact CHAT_B to be online
```
If the base node is required to establish comms between the chat clients, then it should be stopped after the chat clients' online status is confirmed.

brianp replied:
Yes, stopping the base node and having the clients communicate does work, I'll update the test.

hansieodendaal commented on integration_tests/tests/features/Chat.feature:
```diff
@@ -0,0 +1,43 @@
+    Then CHAT_B will have 1 message with CHAT_A
+
+  Scenario: Message counts are distinct
+    Given I have a seed node SEED_A
+    When I have a chat client CHAT_A connected to seed node SEED_A
+    When I have a chat client CHAT_B connected to seed node SEED_A
+    When I have a chat client CHAT_C connected to seed node SEED_A
```
The preferred test mode should be without the base node running

brianp committed:
Stop the base node to ensure comms are going direct

brianp committed:
Move the encrypted send direct into the requester

hansieodendaal reviewed (commented):
Looking good, I just have some comments about the cucumber scenarios.

SWvheerden commented on applications/tari_base_node/log4rs_sample.yml:
```diff
@@ -124,6 +124,11 @@ loggers:
     appenders:
       - network
+  # Route log events sent to the "contacts" logger to the "network" appender
+  contacts:
```
we dont have contact events in the base_node right?

brianp replied:
We shouldn't no. I didn't know how specific each setup file was supposed to be. Shall I take this out?

SWvheerden commented on base_layer/contacts/migrations/2023-03-14-101758_create_messages/up.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE messages (
+    address    BLOB                NOT NULL,
+    message_id INTEGER PRIMARY KEY NOT NULL,
```
rather make this bytes, and create a uuid.

brianp replied:
Yes. Totally forgot about that convo once I was in the weeds. UUID makes a lot more sense from a debugging and linking perspective.

SWvheerden commented on base_layer/contacts/src/contacts_service/handle.rs:
```diff
@@ -224,4 +228,31 @@ impl ContactsServiceHandle {
         }
     }
+
+    pub async fn get_messages(
+        &mut self,
+        pk: TariAddress,
+        x: u64,
+        y: u64,
```
Can you add docs here, and or change the x or y variable names? It's unclear what they do. I am assuming it's paging?

SWvheerden commented on base_layer/contacts/src/contacts_service/service.rs:
```diff
@@ -232,6 +266,47 @@ where T: ContactsBackend + 'static
                 let result = self.get_online_status(&contact).await;
                 Ok(result.map(ContactsServiceResponse::OnlineStatus)?)
             },
+            ContactsServiceRequest::GetMessages(pk, _x, _y) => {
+                let result = self.db.get_messages(pk);
```
I think either drop x and y now and make it get_all_messages
or we choose a type for x,y  and impl
I don't like us having some types un implemented here.

brianp replied:
i'll pull them out for now, we can implement paging later.

SWvheerden commented on base_layer/contacts/src/contacts_service/storage/database.rs:
```diff
@@ -30,39 +30,13 @@ use log::*;
-}
-
 /// This trait defines the functionality that a database backend need to provide for the Contacts Service
 pub trait ContactsBackend: Send + Sync + Clone {
```
this is fine for now, but see: #5320 
This design has been copied over from our badly designed first backend on the base_node which uses a key-value store and not a relational db.

SWvheerden commented on base_layer/contacts/src/contacts_service/storage/types/contacts.rs:
```diff
@@ -0,0 +1,184 @@
+    ) -> Result<ContactSql, ContactsServiceStorageError> {
+        // Note: `get_result` not implemented for SQLite
+        let contact = ContactSql::find_by_node_id(node_id, conn)?;
+        if diesel::delete(contacts::table.filter(contacts::node_id.eq(node_id))).execute(conn)? == 0 {
```
You cant use `.num_rows_affected_or_not_found(1)?;` like above?

brianp replied:
This code was moved over from its previous location in the refactor not written new. I can try out the recommendation though.

SWvheerden commented on base_layer/contacts/src/contacts_service/storage/types/contacts.rs:
```diff
@@ -0,0 +1,184 @@
+#[derive(Clone, Debug, Queryable, Insertable, PartialEq, Eq)]
+#[diesel(table_name = contacts)]
+pub struct ContactSql {
+    pub address: Vec<u8>,
+    node_id: Vec<u8>,
```
Whats the reason we have both here?
address = [node_id][CRC + network]
Asking the node id from address is a simple slice read

SWvheerden commented on base_layer/contacts/src/contacts_service/types/contact.rs:
```diff
@@ -0,0 +1,67 @@
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Contact {
+    pub alias: String,
+    pub address: TariAddress,
+    pub node_id: NodeId,
```
I asked above, but same here, why store both.

brianp replied:
Ahh, not sure. This code was moved in the refactor, not written new for this PR.

SWvheerden commented on comms/dht/src/outbound/requester.rs:
```diff
@@ -52,6 +52,33 @@ impl OutboundMessageRequester {
 
     /// Send directly to a peer. If the peer does not exist in the peer list, a discovery will be initiated.
     pub async fn send_direct<T>(
```
```suggestion
    pub async fn send_direct_encrypted<T>(
```
We are changing the behaviour of an already existing pub API, I would force an error here to force users to look which one to use.

brianp replied:
Like have three functions:
- `send_direct_encrypted`
- `send_driect_unencrypted`

And then `send_direct` would now return an error?

SWvheerden commented on integration_tests/log4rs/wallet.yml:
```diff
@@ -93,6 +93,11 @@ loggers:
     appenders:
       - network
     additive: false
+  contacts:
```
We need to add this to the default log of the wallet as well. 
But I would also use unique appended and not `- network`

SWvheerden reviewed (commented):
This is looking good

brianp committed:
Updates after rebase

brianp committed:
Remove unused paging params

brianp committed:
Update as per review

brianp committed:
Change incremental primary key to uuid

The ids shouldn't increment per client this would be a useless function.
Having a single message between two clients share a common uuid
increases ease of integration for future functionality such as
validation messages have arrived and read receipts etc.

brianp committed:
Remove old function name

Instead of changing the default impl of an existing function, move the
old function to the unencrypted name, make the new function with an
encrypted name, and let any call to the old api fail.

brianp committed:
Ignore prost in contacts when running machete

SWvheerden committed:
Merge branch 'development' into feature-chat-scaffold

SWvheerden commented:
Ack
